### PR TITLE
Remove caching from FreqDist

### PR DESCRIPTION
**Pull request by stevenbird**
This commit removes caching from FreqDist. This needs more testing, and classes that use FreqDist need to be updated. Cf https://github.com/nltk/nltk/issues/86

### CODE REVIEW

**Comment by stevenxxiu** [comment]
I see that the `inc` method is now removed, should the classes be updated too regarding this?

**Comment by stevenxxiu** [comment]
I've modified `SimpleGoodTuringProbDist` in the above commit so that it passes the unit tests in probability.doctest, I don't think it needs caching, as `_r_nr` is only used once on `__init__`, I've also changed some method names in `FreqDist` so they conform to PEP-8.

File: nltk/probability.py [hunk 1]
```diff
@@ -264,9 +208,9 @@ def freq(self, sample):
         :type sample: any
         :rtype: float
         """
-        if self._N == 0:
+        if self.N() == 0:
```
Comment (kmike) [inline nltk/probability.py:211]: Maybe calculate N once and store it in a local variable to match old behavior?

File: nltk/probability.py [hunk 2]
```diff
@@ -490,19 +330,17 @@ def __repr__(self):
         """
         return '<FreqDist with %d samples and %d outcomes>' % (len(self), self.N())
 
-    def __str__(self):
+    def __str__(self, maxlen = 10):
```
Comment (kmike) [inline nltk/probability.py:333]: I think it is better to have a separate method with 'maxlen' parameter and call it from `__str__`

File: nltk/probability.py [hunk 3]
```diff
@@ -199,36 +160,19 @@ def Nr(self, r, bins=None):
...
-        """
-        if self._Nr_cache is None:
-            self._cache_Nr_values()
-        return sorted(self._Nr_cache.items())
-
-    def _cache_Nr_values(self):
-        Nr = defaultdict(int)
         for sample in self:
             c = self.get(sample, 0)
             if c == 0 and c not in Nr:
```
Comment (heatherleaf) [inline nltk/probability.py:171]: `_Nr` not `Nr`

File: nltk/probability.py [hunk 4]
```diff
@@ -186,12 +149,10 @@ def hapaxes(self):
         """
         return [item for item in self if self[item] == 1]
 
-    def Nr(self, r, bins=None):
+    def Nr(self, bins=None):
```
Comment (heatherleaf) [inline nltk/probability.py:152]: Here are two equivalent definitions of `Nr(...)`:

```
_Nr = defaultdict(int)
for count in self.values():
    _Nr[count] += 1
# Special case for Nr[0]:
_Nr[0] = bins - self.B() if bins is not None else 0
```

Or this one, even shorter:

```
_Nr = Counter(self.values())
# Special case for Nr[0]:
_Nr[0] = bins - self.B() if bins is not None else 0
```

The latter is of course much nicer, but it's also 4x slower than the defaultdict version (using timeit on python 3.3).

File: nltk/probability.py (hunk 2 above)
Comment (stevenbird) [inline nltk/probability.py:333]: +1

File: nltk/probability.py (hunk 3 above)
Comment (stevenbird) [inline nltk/probability.py:171]: Thanks

**Comment by stevenbird** [comment]
I've incorporated all these contributions (except caching of N()) into a new `fdist` branch. Further testing welcome.

Note that there's still a problem:
$ python ./probability.py
Traceback (most recent call last):
  ...
  File "./probability.py", line 897, in _calculate_Tr
    Tr[r] += self._heldout_fdist[sample]
IndexError: list index out of range

**Comment by heatherleaf** [comment]
I got rid of that error by letting `Tr` be a defaultdict(float) instead of a list:

```
def _calculate_Tr(self):
    ....
    Tr = defaultdict(float)
    for sample in self._heldout_fdist:
        ...
```

 But then I got this error instead:

```
  File "/Users/peter/Projekt/NLTK/nltk/nltk/probability.py", line 879, in <listcomp>
    Nr = [base_fdist.r_Nr(r, bins) for r in range(self._max_r+1)]
TypeError: r_Nr() takes from 1 to 2 positional arguments but 3 were given
```

So I replace the list comprehension on line 879 with this:

```
    r_Nr = base_fdist.r_Nr(bins)
    Nr = [r_Nr[r] for r in range(self._max_r+1)]
```

 And now I get this error:

```
  File "/Users/peter/Projekt/NLTK/nltk/nltk/probability.py", line 951, in prob
    return self._estimate[r]
IndexError: list index out of range
```

I don't know that much about the HeldoutProbDist, and I don't have the time to continue right now.

**Comment by stevenxxiu** [comment]
Think I got rid of the bug now, was caused by a bug in the `max()` implementation.

**Comment by stevenbird** [comment]
Oh, well done @jskda

**Comment by stevenxxiu** [comment]
Thanks! Forgot to remove a debug statement, fixed that now.

**Comment by stevenbird** [comment]
Thanks. I've manually added the bugfixes to probability.py in the fdist branch at https://github.com/nltk/nltk/tree/fdist. Please take a look. There are still problems (to see them, run "python ./probability.py").

**Comment by heatherleaf** [comment]
Isn't this a really slow way of calculating (line 879 in `HeldoutProbDist.__init__`)? Since `.Nr()` calls `.r_Nr()` for every r:

```
    Nr = [base_fdist.Nr(r, bins) for r in range(self._max_r+1)]
```

How about this instead:

```
    r_Nr = base_fdist.r_Nr(bins)
    Nr = [r_Nr[r] for r in range(self._max_r+1)]
```

**Comment by heatherleaf** [comment]
The problems seem to be very old. I tried to checkout old NLTK versions, and all of them ran into the same problems:
- first a lot of `UserWarning: Probability distribution <SimpleGoodTuringProbDist based on 10 samples> sums to 0.9128893773581336; generate() is returning an arbitrary sample.`
- then there is an `AssertionError: Bins parameter must not be less than freqdist.B() + 1`

**Comment by heatherleaf** [comment]
I first get the AssertionError after commit e5f2a9e, 3 years ago... Has noone tested the probability demo since then?

**Comment by stevenxxiu** [comment]
I think the cause of the assertion error is that in the commit @heatherleaf mentioned, `SimpleGoodTuringProbDist` started using the total number of bins as the third parameter of `__init__` instead of using the "Katz cutoff threshold". I don't think the Katz cutoff exists in the nltk codebase anymore (I did a grep of "katz"), can this portion of the demo be removed?

I've added my suggested fix above.

**Comment by stevenxxiu** [comment]
I've added `SUM_TO_ONE=False` to `SimpleGoodTuringProbDist` to fix the not summing to one issue, in the commit here on my branch: https://github.com/jskda/nltk/commit/5e541f6de892f5c1a97d34eb1456b9dfa64fa823, as I don't think sample probabilities in the simple good turing distribution sum to one.

**Comment by stevenbird** [comment]
@jskda please submit a pull request against the fdist branch in the nltk repository.

**Pull request merged**